[DataGridPremium] Nested Lazy Loading (#21043)
mui/mui-x: docs/data/data-grid/server-side-data/lazy-loading.md

{{"demo": "ServerSideLazyLoadingModeUpdate.js", "bg": "inline"}}

## Nested lazy loading

To use lazy loading for the nested data (tree data or row grouping), pass the `lazyLoading` flag to the Data Grid and the `dataSource` prop to the Data Grid.

> The nested lazy loading doesn't support the "infinite loading" mode, and you must provide the row count upfront for all the parents including the root, this is to avoid the unnecessary flickers during the loading process.

Reviewer: ```suggestion
The nested lazy loading doesn't support the [infinite loading](/x/react-data-grid/server-side-data/lazy-loading/#infinite-loading) mode, and you must provide the row count upfront for all the parents, including the root, this is to avoid the unnecessary flickers during the loading process.
```

### With tree data

The demo below exhibits the behavior of the nested lazy loading with tree data.

> {{"demo": "ServerSideTreeDataNestedLazyLoading.js", "bg": "inline"}}

Reviewer 2: It looks like sorting breaks children expansions – both in tree view and row grouping.
Same with filtering.

https://github.com/user-attachments/assets/deedc920-29f7-4f88-b4ae-a7e634d912ae

Reviewer: Also, on sort, the last column remains in the loading state

<img width="686" height="415" alt="Image" src="https://github.com/user-attachments/assets/b72ecc42-feae-4b0e-a3fc-491d61f08373" />

Probably related to https://github.com/mui/mui-x/pull/21043/changes#r3208536866

Author: Fixed in https://github.com/mui/mui-x/pull/21043/commits/a465aa818c45be79da4091b3d7d91c5b62389cea

### With row grouping [<span class="plan-premium"></span>](/x/introduction/licensing/#premium-plan 'Premium plan')

The demo below exhibits the behavior of the nested lazy loading with row grouping.

{{"demo": "ServerSideRowGroupingNestedLazyLoading.js", "bg": "inline"}}

## Group expansion

Group expansion with nested lazy loading works similarly to how it's described in the [client-side row grouping documentation](/x/react-data-grid/row-grouping/#group-expansion).
The Data Grid expands newly loaded groups based on the `defaultGroupingExpansionDepth` and `isGroupExpandedByDefault()` props, then fetches the visible child skeleton rows for those expanded groups.

> Groups that are loaded later while scrolling use the same behavior.

Reviewer: ```suggestion
Groups that are loaded later while scrolling behave the same.
```

### Tree data [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan 'Pro plan')

The following demo uses `defaultGroupingExpansionDepth={1}` to expand top-level tree data groups as they are loaded.

> {{"demo": "ServerSideLazyLoadingTreeDataGroupExpansion.js", "bg": "inline"}}

Reviewer: Closing down the group does not load the children of the newly visible group. You have to scroll to trigger the update.

https://github.com/user-attachments/assets/400ebbf6-8b25-4211-b876-f0e10bd37ce6

Author: Fixed https://github.com/mui/mui-x/pull/21043/commits/a465aa818c45be79da4091b3d7d91c5b62389cea

### Dynamically updated datasets

The following demos use `dataSourceRevalidateMs="3_000"` (3 seconds) to revalidate the current viewport and get the latest stock prices for the loaded rows.

> In the following demos, note that the row IDs stay stable but the row values change over time.

Reviewer: ```suggestion
Note that the row IDs stay stable but the row values change over time.
```

This can stay as it was

#### Plain data

{{"demo": "ServerSideLazyLoadingRevalidation.js", "bg": "inline"}}

#### Tree data [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan 'Pro plan')

{{"demo": "ServerSideLazyLoadingTreeDataRevalidation.js", "bg": "inline"}}

#### Row grouping [<span class="plan-premium"></span>](/x/introduction/licensing/#premium-plan 'Premium plan')

> {{"demo": "ServerSideLazyLoadingGroupingRevalidation.js", "bg": "inline"}}

Reviewer: WDYT of having these 3 as one demo with 3 grids in 3 tabs?
This way, you can see all variations, without them taking too much space, if this is not something that you are looking for

Author: I personally preferred having separate demos for accessibility pov, but I don't have a strong opinion on this one. Updated: https://github.com/mui/mui-x/pull/21043/commits/0b8b78bc6c526378b7858d6dcd8615b0e55abd62
